### ISSUE TITLE
checkOnChange should be debounced, not throttled

### ISSUE THREAD

**Issue by Arti-Art** [Shopify/theme-check#542]
I would like to address the way the plugin checks the syntax, especially with the checkOnChange setting.
When a character is typed in any liquid file, it will run a theme check on absolutely EVERYTHING in the working directory, and this check will not run again until the current check has finished.
Here are the problems:

- When a character is typed and all subsequent characters are typed before all directories are finished being scanned, there will be no re-scan for those subsequent characters. Ex: If I type "crocodile", the scan starts with the letter "C", and if I finish typing "ROCODILE" before the scan completes, we have done a complete check of all directories only for the letter "C", not the whole word "crocodile. This looks highly inefficient.
- Looks like this issue is the reason that I often get warnings end errors when there should be none, and they disappear only after I save the file (because my checkOnSave is activated). The theme check often runs on incomplete syntax, and often skips the check when what I was trying ty type is complete.
- When a character is typed, ALL of the directories get scanned. Is it necessary to scan everything? My computer is very fast, but 'm sure this process is causing significant performance issues for others. If I am working in a single file for the whole day, I still end up having scanned all of my folders and files several hundreds times by the end of the the day!

So, to recap, theme check runs when it doesn't need to. Theme check does NOT run when it NEEDS to. Theme check constantly scans files that have not changed.

Example video:
[https://www.youtube.com/watch?v=2HRjf2plJws](https://www.youtube.com/watch?v=2HRjf2plJws)

**Comment by vfonic** [comment]
I just came here to report this same issue with full theme check. Good to know it's not full theme check, but you can still see that it's quite slow. It takes 1+ second on my computer, which I consider very slow for "check on change" aka "instant feedback".

The scanning being throttled and not debounced is also an issue I've noticed.

**Comment by charlespwd** [comment]
I think I have something that will work for everyone: 

How about we have another config—`"themeCheck.onlySingleFileChecks"`—that, when true, disables whole theme checking entirely. That means no MissingTranslations, UnusedSnippets, etc. It also means that the problem tab will only ever show the errors for _the files that are opened._

This should speed up execution drastically (a theme with 1k+ errors runs single file check in 0.01ms, the whole theme check takes ~7s) at the small cost of not having _all the errors all of the time_. We could also, maybe, only do the single file checks on change, and somehow debounce the full theme check when we're done. Or only do full theme checks on save.

**Comment by charlespwd** [comment]
Saying this because, surprisingly, debouncing is kind of a gross thing to do in Ruby. It doesn't play nice in a concurrent world. Much easier to do in JS.

**Comment by vfonic** [comment]
> This should speed up execution drastically (a theme with 1k+ errors runs single file check in 0.01ms

😍 
This would be great! :)

This config could live along the other config settings. It's like "focus" mode for your specs. :)
The developers (aka users) can then decide when they'd want to run the full check, maybe as a pre-commit hook and/or on CI, maybe on file save, etc.

I wouldn't make it so that `"themeCheck.onlySingleFileChecks"` implies that full theme check would be run on save.

Maybe existing config options can be changed from boolean to string:

```json
  "themeCheck.checkOnChange": "onlyCurrentFile|openFiles|allFiles",
  "themeCheck.checkOnOpen": "onlyCurrentFile|openFiles|allFiles",
  "themeCheck.checkOnSave": "onlyCurrentFile|openFiles|allFiles",
```

I don't know if all of the options would be useful for all of these settings and how would the migration from boolean to string work for a VS Code extension config.

**Comment by vfonic** [comment]
> I meant this: if the option is off, then we're going to do the fast thing while you type, and the slow thing on save. If onlySingleFileChecks is on, it's always on. (Maybe the cmd+shift+p Run checks command could force run all of them though).

I agree, only currently open file should be checked on change.

I was thinking more about how other code analyzing tools work. Especially whether they analyze more than one file and/or basically anything outside of a single file.

The more I think of it, the more I believe the answer is: No.

If I use Prettier, it will only format the current file on save. If I use ESLint, it will only lint the currently open file. Same if I use Stylelint.

I believe this is an experience that developers are used to. Developers expect this tool to behave similarly.
Can I run ESLint on the whole project? Sure.
Can I run theme-check on the whole project? Yes, but when I want to do so: via CLI command and/or on CI.

**Comment by vfonic** [comment]
It works great! Thank you!

I didn't even have to install dev version of VS Code extension. I just downloaded the changes in this repo (theme-check), built the gem and installed it locally. Then I added the "unknown"/"non-existing" setting option in VS Code settings JSON:

```
"themeCheck.onlySingleFileChecks": true
```

And it worked! :)

**Comment by vfonic** [comment]
Sorry about that! :D I was too excited to give it a try.

I really love this change! 🚀 

Now I can leave theme-check server running in the background, constantly (and immediately) checking the code.

What's next? Are you planning on adding some formatting? Perhaps something that could be set to run on save? Like these things:

```liquid
{%-for product in products-%}
```

=>

```liquid
{%- for product in products -%}
```

**Comment by charlespwd** [comment]
You should already be able to fix those with the quickfix/autofix keybind :)

Quickfix for the popup.
Autofix for the "preferred" fix (usually the one below the cursor).

https://user-images.githubusercontent.com/4990691/155567649-991fb185-f656-4072-8992-85022ae1ecaa.mp4

**Comment by charlespwd** [comment]
As for a formatter, yep it's on my TODO but it's really hard to get right 🙃 and I don't have enough bandwidth to maintain more stuff. We're hiring though so hopefully that will unlock more stuff in the near future :)
